pepper-apis/src/main/java/org/broadinstitute/ddp/export/ExportUtil.java

```diff
@@ -115,4 +119,66 @@ public static void computeActivityAttributesSeen(ActivityInstanceDao instanceDao
                 .findSubstitutionNamesSeenAcrossUsersByActivityAndVersion(activityId, versionId);
         activity.addAttributesSeen(names);
     }
+
+    /**
+     * If address was snapshotted and addressGuid saved to activity_instance_substitution with key = ADDRESS_GUID
+     * then try to get this address. If not found snapshotted address then this method returns a specified defaultAddress.
+     *
+     * @param handle                        jdbi handle
+     * @param activityInstanceSubstitutions substitutions of all activity instances of a current participant:
+     *                                      try to find from it a substitution snapshotted address (we consider
+     *                                      that snapshotted address exist then only one is stored among all activity instances)
+     * @param addressService                service for finding an address bu guid
+     * @param defaultAddress                default address which to return in case of snapshotted address not found
+     * @return MailAddress    detected snapshotted address or defaultAddress (f snapshotted not found)
+     */
+    public static MailAddress getSnapshottedAddress(
+            Handle handle,
+            Map<Long, Map<String, String>> activityInstanceSubstitutions,
+            AddressService addressService,
+            MailAddress defaultAddress) {
+        String addressGuid = null;
+        if (activityInstanceSubstitutions != null) {
+            addressGuid =
+                    activityInstanceSubstitutions.values().stream()
+                            .flatMap(subs -> subs.entrySet().stream())
+                            .filter(subs -> subs.getKey().equals(I18nTemplateConstants.Snapshot.ADDRESS_GUID))
+                            .map(subs -> subs.getValue())
+                            .findFirst()
+                            .orElse(null);
```

Comment:
So we're looking through all the substitutions for all activity instances to grab the snapshotted address? This doesn't look right... What happens if we have mailing address component in two different activities? What happens if we have multiple activity instances of the same activity? Each individual activity instance should have its own snapshotted address, and we shouldn't use another instance's snapshotted address for this instance!

I had an idea while reading through your PR. How does this sound?

(1) We bulk query for "snapshotted, non-default addresses" upfront. We can do this by adding bulk query to our "extract participant dataset" pipeline. It looks like all the various exports should go through `ExportUtil.extractParticipantsFromResultSet()` so that's probably a good place to put it. Try to come up with a bulk query that looks for non-default addresses for a set of participant ids.

* The default address should have already been queried as part of building the `Stream<Participant> resultset` that is passed in by caller. The default address should have been stored in the `Participant` object.
* There should only be one default address, so the rest are non-default addresses. And we know that non-default addresses are what we use for "snapshotting".
* We can filter out the default address in our bulk query by looking in the `default_mailing_address` and excluding that `address_id` from our result set.
* You have the `Stream<Participant> resultset` and we're already iterating through it, so it is ideal to collect all the participant ids from the `resultset` into a `Set`, so we can do a bulk query for all these participants in one go.
* After we done the bulk query, we can iterate through the queried rows, group them by participant, and assign them to the appropriate participant (notice that we have `Map<String, Participant> participants` which should help with this! But you might need to change how we map things from participant guid `String` to participant id `Long` for an easier implementation).
* There's probably more efficient ways to do this, but the end result is that we have bulk  queried all non-default addresses for all participants in our dataset.

(2) In `Participant` class, we have the activity instance substitutions. We are getting passed the list of snapshotted addresses from (1). So at this point, we can do a bit of preprocessing to figure out which activity instance has which snapshotted address.
* Perhaps we add a method called `associateInstanceWithSnapshottedAddresses(Map<String, MailAddress> nonDefaultAddresses)` (need a better name here...) which step (1) will call.
* The end result is that in `Participant` we store a mapping of `Map<Long, MailAddress> instanceIdToNonDefaultAddress` (any better name?)
* Now, the rest of the export code don't need to grab a `Handle` object and don't need to run any more queries! It's all there in the extracted participant dataset!

(3) For exports that use `ComponentDataSupplier`, we can use the supplier as an "interface" to get the snapshotted address.
* Pass in the `instanceIdToAddress` mapping  to `ComponentDataSupplier` constructor.
* Either add a new method or update existing `getAddress()` method to take a `long activityInstanceId` argument.
* Now, getting an address means looking in the map. If there is an entry, use that, otherwise fallback to the default address.
* Tip: in `ActivityResponseCollector` you can pass the `instance` to `collectComponentIntoRecord()`.

(4) For exports that don't use `ComponentDataSupplier` but has access to the `Participant` object, we can do similar thing and add method to `Participant` to fetch an address for an activity instance (which will look it up in the map or use default).

By doing a bulk query, things are bit more efficient. And by collecting all the data upfront, now we don't need to open up a transaction while exporting data and holding the transaction open for so long.

Reply:
Implemented this logic